### PR TITLE
Use the CommonConfiguration structure for roles.

### PR DESCRIPTION
**Pull request by razvan**
# Description

Part of #273 

Refactor the SparkApplication CRD to be more in line with other operators.

The three main roles (job, driver and executor) are now concrete `RoleGroup` objects under an `Role` named "default" that is constructed by the operator internally.

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
# Author
- [x] Changes are OpenShift compatible
- [ ] CRD changes approved
- [x] Helm chart can be installed and deployed operator works
- [x] Integration tests passed (for non trivial changes)
```

```[tasklist]
# Reviewer
- [x] Code contains useful comments
- [x] Documentation added or updated
- [x] Changelog updated
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
```

### CODE REVIEW

**Comment by razvan** [comment]
CI https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/spark-k8s-operator-it-custom/101/

File: rust/crd/src/lib.rs [hunk 1]
```diff
@@ -863,195 +895,9 @@ pub struct NodeSelector {
 
 impl Atomic for NodeSelector {}
```
Comment (fhennig) [inline rust/crd/src/lib.rs:896]: What does the NodeSelector do? I can't seem to find any references to it in the code. Also, why is it Atomic?

File: rust/crd/src/roles.rs [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+}
+
+// TODO: remove this when switch to pod overrides ???
+#[derive(Clone, Debug, Default, Deserialize, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct VolumeMounts {
+    pub volume_mounts: Option<Vec<VolumeMount>>,
+}
+
+impl Atomic for VolumeMounts {}
```
Comment (fhennig) [inline rust/crd/src/roles.rs:382]: Why are the VolumeMounts Atomic?

File: rust/operator-binary/src/main.rs [hunk 3]
```diff
@@ -108,13 +115,16 @@ async fn main() -> anyhow::Result<()> {
...
                 pod_driver_controller::error_policy,
-                Arc::new(pod_driver_controller::Ctx {
-                    client: client.clone(),
-                }),
+                Arc::new(client.clone()),
             )
             .map(|res| report_controller_reconciled(&client, &format!("{OPERATOR_NAME}.{POD_DRIVER_CONTROLLER_NAME}"), &res))
             .instrument(info_span!("pod_driver_controller"));
 
+            // Create new object because Ctx cannot be cloned
```
Comment (fhennig) [inline rust/operator-binary/src/main.rs:123]: Why can't clone not be implemented?

Could there maybe be a function on the Ctx for a pseudo clone?

**Review by fhennig — CHANGES_REQUESTED** [review]
As mentioned before, docs for the roles.rs file would be great, and some of the tests don't look right.

File: docs/modules/spark-k8s/examples/example-sparkapp-configmap.yaml [hunk 4]
```diff
@@ -20,11 +20,13 @@ spec:
...
+      volumeMounts:
+        - name: cm-job-arguments # <6>
+          mountPath: /arguments  # <7>
   executor:
-    instances: 3
-    volumeMounts:
-      - name: cm-job-arguments # <6>
-        mountPath: /arguments # <7>
+    replicas: 3
+    config:
```
Comment (fhennig) [inline docs/modules/spark-k8s/examples/example-sparkapp-configmap.yaml:29]:
```suggestion
    config:
      replicas: 3
```

File: docs/modules/spark-k8s/examples/example-sparkapp-external-dependencies.yaml [hunk 5]
```diff
@@ -24,11 +24,13 @@ spec:
...
+      volumeMounts:
+        - name: job-deps
+          mountPath: /dependencies  # <6>
   executor:
-    instances: 3
-    volumeMounts:
-      - name: job-deps
-        mountPath: /dependencies  # <6>
+    replicas: 3
+    config:
```
Comment (fhennig) [inline docs/modules/spark-k8s/examples/example-sparkapp-external-dependencies.yaml:33]:
```suggestion
    config:
      replicas: 3
```

File: docs/modules/spark-k8s/examples/example-sparkapp-image.yaml [hunk 6]
```diff
@@ -19,24 +19,27 @@ spec:
...
   executor:
-    instances: 3
-    resources:
-      cpu:
-        min: "1"
-        max: "4"
-      memory:
-        limit: "2Gi"
+    replicas: 3
+    config:
```
Comment (fhennig) [inline docs/modules/spark-k8s/examples/example-sparkapp-image.yaml:39]:
```suggestion
    config:
      replicas: 3
```

File: docs/modules/spark-k8s/examples/example-sparkapp-pvc.yaml [hunk 7]
```diff
@@ -22,11 +22,13 @@ spec:
...
+      volumeMounts:
+        - name: job-deps
+          mountPath: /dependencies # <5>
   executor:
-    instances: 3
-    volumeMounts:
-      - name: job-deps
-        mountPath: /dependencies # <5>
+    replicas: 3
+    config:
```
Comment (fhennig) [inline docs/modules/spark-k8s/examples/example-sparkapp-pvc.yaml:31]:
```suggestion
    config:
      replicas: 3
```

File: docs/modules/spark-k8s/examples/example-sparkapp-s3-private.yaml [hunk 8]
```diff
@@ -22,4 +22,4 @@ spec:
     spark.driver.extraClassPath: "/dependencies/jars/hadoop-aws-3.2.0.jar:/dependencies/jars/aws-java-sdk-bundle-1.11.375.jar"
     spark.executor.extraClassPath: "/dependencies/jars/hadoop-aws-3.2.0.jar:/dependencies/jars/aws-java-sdk-bundle-1.11.375.jar"
   executor:
-    instances: 3
+    replicas: 3
```
Comment (fhennig) [inline docs/modules/spark-k8s/examples/example-sparkapp-s3-private.yaml:25]:
```suggestion
    config:
      replicas: 3
```

File: examples/ny-tlc-report-external-dependencies.yaml [hunk 9]
```diff
@@ -30,11 +30,13 @@ spec:
...
+      volumeMounts:
+        - name: job-deps
+          mountPath: /dependencies
   executor:
-    instances: 3
-    volumeMounts:
-      - name: job-deps
-        mountPath: /dependencies
+    replicas: 3
+    config:
```
Comment (fhennig) [inline examples/ny-tlc-report-external-dependencies.yaml:39]:
```suggestion
    config:
      replicas: 3
```

File: examples/ny-tlc-report-image.yaml [hunk 10]
```diff
@@ -25,4 +25,4 @@ spec:
   sparkConf:
     spark.hadoop.fs.s3a.aws.credentials.provider: "org.apache.hadoop.fs.s3a.AnonymousAWSCredentialsProvider"
   executor:
-    instances: 3
+    replicas: 3
```
Comment (fhennig) [inline examples/ny-tlc-report-image.yaml:28]:
```suggestion
    config:
      replicas: 3
```

File: examples/ny-tlc-report.yaml [hunk 11]
```diff
@@ -32,11 +32,13 @@ spec:
...
+      volumeMounts:
+        - name: cm-job-arguments
+          mountPath: /arguments
   executor:
-    instances: 3
-    volumeMounts:
-      - name: cm-job-arguments
-        mountPath: /arguments
+    replicas: 3
+    config:
```
Comment (fhennig) [inline examples/ny-tlc-report.yaml:41]:
```suggestion
    config:
      replicas: 3
```

**Review by fhennig — CHANGES_REQUESTED** [review]

File: rust/operator-binary/src/main.rs (hunk 3 above)
Comment (razvan) [reply rust/operator-binary/src/main.rs:123]: It's an unfortunate omission that `ProductConfigManager` does not derive Clone and the `config` field is private. It should be fixed in operator-rs but I guess there was no need for it.

File: rust/crd/src/lib.rs (hunk 1 above)
Comment (razvan) [reply rust/crd/src/lib.rs:896]: It's a left over. The functionality is now covered by the affinites.

**Review by fhennig — DISMISSED** [review]
LGTM!

**Comment by razvan** [comment]
CI: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/spark-k8s-operator-it-custom/103/

**Review by fhennig — APPROVED** [review]

File: examples/ny-tlc-report-external-dependencies.yaml [hunk 12]
```diff
@@ -30,11 +30,13 @@ spec:
...
+      volumeMounts:
+        - name: job-deps
+          mountPath: /dependencies
   executor:
-    instances: 3
-    volumeMounts:
-      - name: job-deps
-        mountPath: /dependencies
+    config:
+      replicas: 3
```
Comment (adwk67) [inline examples/ny-tlc-report-external-dependencies.yaml:39]: Replicas should be at same level as config.

File: examples/ny-tlc-report-image.yaml [hunk 13]
```diff
@@ -25,4 +25,5 @@ spec:
   sparkConf:
     spark.hadoop.fs.s3a.aws.credentials.provider: "org.apache.hadoop.fs.s3a.AnonymousAWSCredentialsProvider"
   executor:
-    instances: 3
+    config:
+      replicas: 3
```
Comment (adwk67) [inline examples/ny-tlc-report-image.yaml:29]: Replicas should be at same level as config.

File: examples/ny-tlc-report.yaml [hunk 14]
```diff
@@ -32,11 +32,13 @@ spec:
...
+      volumeMounts:
+        - name: cm-job-arguments
+          mountPath: /arguments
   executor:
-    instances: 3
-    volumeMounts:
-      - name: cm-job-arguments
-        mountPath: /arguments
+    config:
+      replicas: 3
```
Comment (adwk67) [inline examples/ny-tlc-report.yaml:41]: Replicas should be at same level as config.

**Review by adwk67 — CHANGES_REQUESTED** [review]
Just a couple of minor things.

File: examples/ny-tlc-report-external-dependencies.yaml (hunk 12 above)
Comment (razvan) [reply examples/ny-tlc-report-external-dependencies.yaml:39]: We agreed that `replicas` is outside of `config`.

**Review by adwk67 — APPROVED** [review]
LGTM!

**Pull request merged**